src/app/InteractionModelEngine.cpp

```diff
@@ -485,6 +540,217 @@ void InteractionModelEngine::AddReadClient(ReadClient * apReadClient)
+    int32_t usedEventPaths     = static_cast<int32_t>(mEventPathPool.Allocated());
+    int32_t usedReadHandlers   = static_cast<int32_t>(mReadHandlers.Allocated());
+
+    if (usedAttributePaths + aRequestedAttributePathCount <= attributePathCap &&
+        usedEventPaths + aRequestedEventPathCount <= eventPathCap && usedReadHandlers < readHandlerCap)
+    {
+        // We have enough resources, then we serve the requests in a best-effort manner.
+        return true;
+    }
+
+    if ((aRequestedAttributePathCount > static_cast<int32_t>(kMinSupportedPathPerSubscription) &&
+         usedAttributePaths + aRequestedAttributePathCount > attributePathCap) ||
+        (aRequestedEventPathCount > static_cast<int32_t>(kMinSupportedPathPerSubscription) &&
+         usedEventPaths + aRequestedEventPathCount > eventPathCap))
+    {
+        // We cannot offer enough resources, and the subscription is requesting more than the spec limit.
+        return false;
+    }
+
+    bool didEvictHandler = true;
+
+    const auto evictAndUpdateResourceUsage = [&](FabricIndex fabricIndex, bool forceEvict) {
+        bool ret           = TrimFabric(fabricIndex, forceEvict);
+        usedAttributePaths = static_cast<int32_t>(mAttributePathPool.Allocated());
+        usedEventPaths     = static_cast<int32_t>(mEventPathPool.Allocated());
+        usedReadHandlers   = static_cast<int32_t>(mReadHandlers.Allocated());
+        return ret;
+    };
+
+    //
+    // At this point, we have an inbound request that respects minimas but we still don't have enough resources to handle it. Which
+    // means that we definitely have handlers on existing fabrics that are over limits and need to evict at least one of them to
+    // make space.
+    //
+    while ((usedAttributePaths + aRequestedAttributePathCount > attributePathCap ||
+            usedEventPaths + aRequestedEventPathCount > eventPathCap || usedReadHandlers >= readHandlerCap) &&
+           didEvictHandler)
+    {
+        // There might be cases that one fabric has lots of subscriptions with one interested path, while the other fabrics are not
+        // using excess resources. So we need to do this multiple times until we have enough room or no fabrics are using excess
+        // resources.
+        didEvictHandler = false;
+        for (const auto & fabric : *mpFabricTable)
+        {
+            didEvictHandler = didEvictHandler || evictAndUpdateResourceUsage(fabric.GetFabricIndex(), false);
+            if (usedAttributePaths + aRequestedAttributePathCount <= attributePathCap &&
+                usedEventPaths + aRequestedEventPathCount <= eventPathCap && usedReadHandlers < readHandlerCap)
+            {
+                break;
+            }
+        }
+    }
+
+    // We either have enough resources, or the resource usage from all fabrics are within the quota, so evict one (possibly the
+    // oldest) subscription from the current fabric.
```

Comment:
This is not just evicting one. It's evicting subscriptions one by one in the **accessing** (that should be the term we use here) fabric till we have space.